### ISSUE TITLE
Look int prompt-toolkit as a readline replacement.

### ISSUE THREAD

**Issue by scopatz** [scopatz/xonsh#214]
https://github.com/jonathanslenders/python-prompt-toolkit

It would be awesome if this was a drop-in.

**Comment by wrywerytwreywery** [comment]
This seems like it would be very nice!  In particular, I think this would let us do syntax highlighting with pygments, would allow some more control over tab-completion (to help with improving quoted strings, for example), and might also allow us to do zsh-style tab-completion (where the suggestions happen in place, rather than re-printing each time)...

**Comment by anula** [comment]
Just a question: are you working on this right now, or is it also "help wanted"?

**Comment by anula** [comment]
I had some time today to play with this, so if no one works on this yet here are my findings, otherwise just ignore me.

So I tried to replace all readline+cmd stuff with prompt-toolkit. It required really small amount of work to get something working (if you don't count time needed to learn prompt-toolkit), and here is an example of what I changed:
https://github.com/anula/xonsh/commit/a3176fd0cfc2d443e6b871d6275cb726928b7218
I basically just got rid of readline and cmd replacing it with prompt-toolkit and added syntax highlighting (it is really easy with the library).
Things that don't yet work in my example are:
- completion - it requires implementation of a bit different interface for Completer, but it should not be a problem to add this
- Shift-Tab for indentation
- and there is a not-so-pretty workaround for problem with colorful prompt (that I describe below) which is just printing it using print() before call to get_input()
But implementing those should be just a matter of time and, as @wrywerytwreywery said, their completion systems seems to be a bit more powerful.
The example is also a bit inelegant, I wanted to change as little as possible so that it is easy to see what is added, and because of it there is a lot of unused code, at least one but, etc.

Problems that I've encountered:
- when you pass to get_input() prompt string that contains color escape codes you don't get pretty colorful prompt but rather something like "^A^[[1;34m^B$^A^[[0m^B". Pretty colorful prompt can be achieved differently, though, by using Pygments's styles. I think it would not be really hard to convert format like this: "{BOLD_GREEN}{user}@{hostname}{BOLD_BLUE} {cwd}{BOLD_RED}{curr_branch} {BOLD_BLUE}${NO_COLOR}" accordingly but I am not sure how to give users more freedom in choosing their prompt (I think that now they can just write arbitrary function that returns prompt?). Plus side is, if I understand this correctly, that it should work for Windows in the same way as it works for Linux/Mac OS.
No more real problems for now, but surely it doesn't mean that they are not there.

Now let's move on to nice things:
Generally, I have feeling that design of this toolkit is much more suitable for xonsh than readline is. It lets you plug in more things in more elegant way. 
    - For example: code highlighting - one can even write lexer just for xonsh that will color properly python as well as bash commands.
   - Another example: a history object. You just pass any implementation of 'History' interface, so that would probably help a bit with implementation of richer history. 
   - Another: multiline commands. I am not yet exactly sure how to force it to switch into multiline mode when previous line indicates that it should, but ptpython does it so there must be a way.

There are of course many more possibilities. Question is probably if it is worth to add another dependency and how efficient it would be with code completion.

So that is what I have found out, let me know if you think this is worth it to continue with the migration, and if you want me to work on this.  Even if it doesn't get merged eventually, I'd be happy to.

**Comment by wrywerytwreywery** [comment]
Woo-hoo, thanks for giving this a go!

I think we will need to have something closer to a finished product before we can make a real decision about whether this is a win or not, but my personal feeling is that it is worth it to try to run with this, and the benefits I see (syntax highlighting, more control over tab-completion) outweigh the downside of the added dependency.  I am happy to help how I can (maybe with tab completion?).

> code highlighting - one can even write lexer just for xonsh that will color properly python as well as bash commands.

@scopatz already wrote such a thing for the docs, so we should be able to drop that in!  It isn't perfect (see #173 for some of the limitations), but it should be pretty good.
`from xonsh.pyghooks import XonshLexer`

> and there is a not-so-pretty workaround for problem with colorful prompt (that I describe below) which is just printing it using print() before call to get_input()

I am still getting some garbage characters for my (colorful) prompt even though the code is just using print.  Not sure what is going on there.  Printing the color escape sequences directly seems to work, as with the following:
`print('\x1b[0;31m hello')`

> multiline commands. I am not yet exactly sure how to force it to switch into multiline mode when previous line indicates that it should, but ptpython does it so there must be a way.

`ptpython` looks like it has a pretty complicated setup for this, but one whose style we probably want to mimic.  Instead of calling `get_input` directly, it sets up an instance of `CommandLineInterface` (see [here](https://github.com/jonathanslenders/ptpython/blob/master/ptpython/python_input.py#L106-114)) and has a check to see whether the buffer in question is (currently?) multi-line [further down the file](https://github.com/jonathanslenders/ptpython/blob/master/ptpython/python_input.py#L190-203).  I am not sure how all the pieces fit together, but that might be the place to start looking.

Another option, though, would be just to keep doing what we have been, since that should still work despite the change.

**Comment by anula** [comment]
> @scopatz already wrote such a thing for the docs, so we should be able to drop that in! It isn't perfect (see #173 for some of the limitations), but it should be pretty good.
> from xonsh.pyghooks import XonshLexer

Hah, nice that this is already here!

> I think we will need to have something closer to a finished product before we can make a real decision about whether this is a win or not

Great! Then I will try to implement something "more finished".  For starters, I will try to copy current behavior (so just rewrite Completer and create a simple History object).

> I am still getting some garbage characters for my (colorful) prompt even though the code is just using print.

About prompt, can you tell me if [this page](http://xonsh.org/tutorial.html#customizing-the-prompt) covers everything that user can do to customize it? 
If so, then I guess I can write a conversion to suitable format (create [custom prompt](https://github.com/jonathanslenders/python-prompt-toolkit/blob/ff5cb568fac30ae7b64fef596ffe7f56f6b2d0f6/prompt_toolkit/layout/prompt.py), with some [custom styles](https://github.com/jonathanslenders/ptpython/blob/332ae4d05d535b5271b38233f3f8856dc83e398c/ptpython/style.py) - not yet sure exactly how), but that will require a bit more researching. I will start with just asking on prompt-toolkit issues why color codes don't work.

> ptpython looks like it has a pretty complicated setup for this, but one whose style we probably want to mimic. Instead of calling get_input directly, it sets up an instance of CommandLineInterface (see here) and has a check to see whether the buffer in question is (currently?) multi-line further down the file. I am not sure how all the pieces fit together, but that might be the place to start looking.

So `get_input` is just a shortcut for `CommandLineInterface` with some default values, but I am not yet sure how "buffers" work here, that is probably important. I think I will get to it as soon as I manage to make it work in the same way as it works now.

> Another option, though, would be just to keep doing what we have been, since that should still work despite the change.

Indeed it works, but I guess it would be nice to have multiline support. History object would get whole commands then, that may come in handy.

So, my plan right now is to add simple history object (so simple that it should be easy to merge with #213 if there is such need), and rewrite Completer.
Next step is to add multiline support and prompt, since the former probably won't work correctly without the latter.

**Comment by wrywerytwreywery** [comment]
That all sounds good to me.  Thanks!

> About prompt, can you tell me if this page covers everything that user can do to customize it? 

Yes, I believe that should be it.

It seems like fixing the colors will be hard to implement with pygments.  I think it is worth looking into how `prompt_toolkit` is doing its thing with regard to ASNI color codes in the prompt.

> Indeed it works, but I guess it would be nice to have multiline support. History object would get whole commands then, that may come in handy.

Ahh, yes; that would be very nice.  I guess I have gotten used to hitting the up arrow multiple times to get back the multiline input...

**Comment by anula** [comment]
`FileHistory` in prompt-toolkit has a bit different behavior than history provided by readline. The latter writes all lines of history to the history file at the end of session (similarly to bash) and the former opens history file for each command that is entered and writes it down immediately (but if you press up you don't see commands that were entered in other terminal that is running simultaneously).
Readline has also this problem, that if you open two terminals, enter some commands in first one, close it, then enter some commands in second one and close it as well, history from the first one is overridden. 

I'd personally prefer prompt-toolkit solution. The only problem might be that it reads whole history file to memory and keeps it there throughout the time shell is running. But it should not be much if there is any reasonable limit on lines of history set, and if richer history is implemented it will be rewritten anyway.

Which one is preferred for xonsh? I can also implement any hybrids of those two.

**Comment by anula** [comment]
On second thought, writing each line to file immediately isn't a very good idea if we want to keep maximal limit of lines. Writing at the end of session seems to be more reasonable and still it can be written in a way that prevents overriding.

**Comment by scopatz** [comment]
> On second thought, writing each line to file immediately isn't a very good idea if we want to keep maximal limit of lines.

I agree.  I think we are going to do our own thing with history very soon anyway (as you mentioned)

**Comment by anula** [comment]
Just a quick update: it seems to be easy to get completion to work, but I still didn't manage to get colorful prompt to work. This library assumes that it knows everything about what is written on screen, so using my workaround with `print` doesn't seem like a good idea in the long run (things like this start happening:
![completions_in_wrong_place](https://cloud.githubusercontent.com/assets/4399755/7511831/4838e2f0-f4a8-11e4-821b-f807c41c9909.png) )
I went into a few dead-ends tying to make it work, but I have a few more ideas - I will let you know on Friday or Saturday if anything works.

**Comment by wrywerytwreywery** [comment]
Thanks again for looking into this!  Glad to hear that completion seems to be working.

I think you're right that the `print` solution isn't the right thing in the long run.  It does look like there is a way to force colorization in prompt_toolkit by setting the handler in `DefaultPrompt` (this will probably involve doing some more setup rather than just using `get_input` directly) to return tokens that color the input.  This might require some changes to the way we handle prompts within xonsh, but it feels like it should be possible to translate our current form (string with formatting directives) to the form they would need (function that generates pygments tokens).  I would be happy to put in some work on this if that would be helpful.

**Comment by anula** [comment]
So I'd say that it should be possible to create one Token for each color and then to convert strings that have those colors like `{GREEN}` or even ANSI colors into list of pairs of tokens and text. However, I'd say that it may break some of currently working prompts that rely more on ANSI escape sequences - I am not sure if it is easy (or even possible) to emulate its behavior completely.
I am now looking into how to define new Layout in prompt_toolkit - maybe it is possible to omit the step that causes prompt to display incorrectly if you hook it deeply enough. I don't have time today to research it further, but I will try to do it tomorrow.

On the other hand, maybe it would be more portable solution if we force users to define prompt with markdown that is more high-level than ANSI sequences and then just parsed into pygments style?
It should be possible to let them define their own styles for their own tokens and it seems that it gives [a lot of options](http://pygments.org/docs/styles/#style-rules) to customize "prettiness".

**Comment by scopatz** [comment]
Hi @anula, first off thanks *a ton* for investigating this.  

Also, I just wanted to say that I currently lean towards this being optional (ie the default prompt-toolkit is around, and falling back to readline if it isn't). That way we can retain only one strict dependency (PLY).  I wanted to get your and @wrywerytwreywery's thoughts though.  Thanks again!

**Comment by anula** [comment]
With my current knowledge I'd say it seems like a doable and sensible idea. Surely readline doesn't offer all the possibilities that prompt_toolkit offers, so some features would work only when prompt_toolkit is available.
The only problem that I see right now is prompt. The easiest solution for it to work in both modes would be to just make prompt_toolkit display it correctly, so that is what I am now looking into.

**Comment by scopatz** [comment]
Sounds good!

**Comment by anula** [comment]
Hah, I think I found out why just passing ANSI colors to `get_input` doesn't work. They use [`Char`](https://github.com/jonathanslenders/python-prompt-toolkit/blob/master/prompt_toolkit/layout/screen.py#L31) to represent single character, and that class just manually replaces all color sequences with something else.
The crucial moment is [here](https://github.com/jonathanslenders/python-prompt-toolkit/blob/8c51def7fadbf93319213ee0432695808a845543/prompt_toolkit/layout/screen.py#L133) - they iterate over a text that is to be displayed (the text is separated into smaller parts, each has its own Token attached  - the Token can be connected to a style information), char by char. For each char they create an instance of `Char` (here escape sequences are replaced by something else), and put the char into a two dimensional array that directly represents terminal output. The array is later compared to what was written to screen last time, and things that differ are redrawn. Because they support unicode characters, they compute width of each char using wcwidth library (corresponding function is [here](https://github.com/jonathanslenders/python-prompt-toolkit/blob/aa360ba8d1d4d671c23de7afd75ba549b690bdd8/prompt_toolkit/utils.py#L57)) so if a character takes more than one cell they can react properly.
Because of that, it is really hard to make escape sequences work. Even if one deletes the conversion that happens while creating a `Char`, an escape sequence still have no width so how can it be put into the buffer? To change this behavior one would need to completely rewrite this part of the library.

I managed to create a workaround, but it is "a bit" tricky. Instead of passing a string as a prompt, I pass a list of chars. Only some of those chars are actually not single chars but chars together with escape sequences. Those FakeChars are created in a way that makes [this function](https://github.com/jonathanslenders/python-prompt-toolkit/blob/aa360ba8d1d4d671c23de7afd75ba549b690bdd8/prompt_toolkit/utils.py#L57) return width equal to width of what will be written on screen, and `Char` doesn't seem to mind that it keeps a bit more than one char.
[Here](https://gist.github.com/anula/5748a896cb0c749257a5#file-gistfile1-py-L250) is sample implementation (it is just a single file that can be run on its own). It is a bit long, but the whole beginning really doesn't matter (I just need to create my own `CommandLineInterface` instance so I copied code from [shortcuts](https://github.com/jonathanslenders/python-prompt-toolkit/blob/193d941c06f12e795ef676e2d064b82498fda983/prompt_toolkit/shortcuts.py) and made my changes).
The real workaround is just `FakeChar` class, and the fact that instead of prompt I pass a list. It should be easy to convert any string containing escape sequences to this form.

The problems with this workaround are probably obvious. It depends on fact that [here](https://github.com/jonathanslenders/python-prompt-toolkit/blob/master/prompt_toolkit/utils.py#L65) when string has length different that one they iterate over it. And generally on internal implementation that **probably** will not be changed.

Unfortunately, I really don't see a way to make it work without depending a bit on internals of the library.

What are your thoughts? Maybe that is an acceptable workaround for some time, and later it would be better to just change the way that users can customize prompt? Or is it just assuming too much?

**Comment by scopatz** [comment]
I think that this is an acceptable workaround for the time being.  We do much worse things to get readline working properly.  I'd also bring up this use case with them once the workaround is implemented and see what they have to say. They will probably have some ideas about what the best way to support us long term are.

**Comment by anula** [comment]
Great, I will proceed with this then.

**Comment by rbrewer123** [comment]
@scopatz, are you comfortable closing this issue considering how far along prompt-toolkit is now?

**Comment by scopatz** [comment]
Yep, @anula has done a great job here and we've even had some support beyond this. Let's call this done for now :)
